src/StreamFactory.php

```diff
@@ -35,9 +36,15 @@ public function createStreamFromFile(string $file, string $mode = 'r'): StreamIn
 
     /**
      * {@inheritDoc}
+     *
+     * @param resource|GdImage $resource
      */
```

Comment:
I am not sure if this is good. Expanding parameter types is considered type safe. The problem is anyone needing `GdImage` support can no longer depend on PSR interface because that would not be type safe.

Wouldn't that go against interoperability goal between libraries since this will be creating dependency on diactoros and this implementation?

Reply:
`GdImage` is considered an _opaque resource type_. The general assumption of PSR-7 and PSR-17's stream support is that it can work with _resources_ of any type. However, _both_ pre-date the opaque resource types.

This suggests two things to me:

- They should allow usage with opaque resource types _where they make sense_. Since you can have image resources spit out content, I'd argue they make sense here, and, in some ways, are better than normal file-based resources where you lose idempotency.
- Alternately, PSR-7 and/or PSR-17 need updates to either explicitly allow or disallow usage with opaque resource types.

For now, I don't think this patch violates interoperability, as PSR-7 and PSR-17 _do_ talk about resources, and this is a type of resource.

Reply:
I would not support `GdImage` via `StreamFactory` TBH

PSR is clear here:
https://github.com/php-fig/http-factory/blob/e616d01114759c4c489f93b099585439f795fe35/src/StreamFactoryInterface.php#L44

If some1 wants to use `GdImage` with `StreamFactoryInterface` they are already lost. If they hardly depend on laminas as an implementation, they don't need PSR and thus can directly instantiate `ImageStream` rather than using the interchangeable factory.  The whole point of using the `StreamFactory` is to provide interchangeability and if we now decide to support `GdImage` that would only fork for laminas but not for guzzle for example. Having these kind of incompatibility is super annoying and thus, I'd say we should not allow that at all.